src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/W3CPropagator.cs

```diff
@@ -0,0 +1,741 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace System.Diagnostics
+{
+    internal sealed class W3CPropagator : DistributedContextPropagator
+    {
+        internal static DistributedContextPropagator Instance { get; } = new W3CPropagator();
+
+        private const int MaxBaggageEntriesToEmit = 64;     // Suggested by W3C specs
+        private const int MaxBaggageEncodedLength = 8192;   // Suggested by W3C specs
+        private const int MaxTraceStateEncodedLength = 256; // Suggested by W3C specs
+
+        private const char Equal = '=';
+        private const char Percent = '%';
+        private const char Replacement = '\uFFFD'; // �
+
+        public override IReadOnlyCollection<string> Fields { get; } = new ReadOnlyCollection<string>(new[] { TraceParent, TraceState, Baggage });
+
+        public override void Inject(Activity? activity, object? carrier, PropagatorSetterCallback? setter)
+        {
+            if (activity is null || setter is null || activity.IdFormat != ActivityIdFormat.W3C)
+            {
+                return;
+            }
+
+            string? id = activity.Id;
+            if (id is null)
+            {
+                return;
+            }
+
+            setter(carrier, TraceParent, id);
+            if (!string.IsNullOrEmpty(activity.TraceStateString))
+            {
+                InjectTraceState(activity.TraceStateString, carrier, setter);
+            }
+
+            InjectW3CBaggage(carrier, activity.Baggage, setter);
```

Comment:
Can we not inject activity.Baggage in the new version? I.e. it will only ever go to Correlation-Context and not adding it to W3C baggage would discourage people from using it (which is a great thing).

We can add `InjectBaggage(IEnumerable<KeyValuePair<string, string?>>, ...)` making it symmetrical with `ExtractBaggage` one and decoupled from activity context

Reply:
Can we table this for discussion as I don't think we can resolve it on spot here? We still have the chance to change it later if we agreed on. Thanks!

Reply:
We discussed this offline and decided to keep emitting activity baggage in the new propagator.